src/Pate/Monad.hs

```diff
@@ -459,19 +484,27 @@ applyAssumptionSet asm f = withSym $ \sym -> do
     doRebind = rebindWithFrame' sym cache asm
   liftIO $ PEM.mapExpr sym doRebind f
 
-
 -- | First check if an assumption is satisfiable before assuming it. If it is not
 -- satisfiable, return Nothing.
 withSatAssumption ::
   HasCallStack =>
-  PT.Timeout ->
-  W4.Pred sym ->
+  AssumptionSet sym v ->
   EquivM sym arch f ->
   EquivM sym arch (Maybe f)
-withSatAssumption timeout asm f = do
-  isPredSat timeout asm >>= \case
-    True ->  Just <$> (withAssumption asm f)
-    False -> return Nothing
+withSatAssumption asm f = withSym $ \sym -> withSolverProcess $ \sp -> do
+  p <- liftIO $ getAssumedPred sym asm
+  case W4.asConstantPred p of
+    Just False -> return Nothing
+    Just True -> Just <$> f
+    _ -> do
+      curAsm <- currentAsm
+      CMR.local (\env -> env { envCurrentFrame = Some (asm <> curAsm) }) $ do
+        IO.withRunInIO $ \inIO -> WPO.inNewFrame sp $ do
+          W4.assume (WPO.solverConn sp) p
+          b <- WPO.checkAndGetModel sp "check assumptions" >>= asSat
```

Comment:
Is there some argument we can write down for why we are not concerned that we haven't specified a timeout here? Is this always easy enough to succeed? Can we not continue if we don't have an accurate answer?

Reply:
I've got a patch that addresses this that will be committed later. Specifically I had to put in some fixes to get timeouts working again: https://github.com/GaloisInc/pate/issues/316

I'll put a comment here.